### ISSUE TITLE
[consistent-type-imports] prefer inline type import syntax

### ISSUE THREAD

**Issue by akphi** [typescript-eslint/typescript-eslint#4338]
### Description

I would like to propose an extension to the existing `consistent-type-imports` rule. This is linked to the work done in https://github.com/typescript-eslint/typescript-eslint/pull/4237

I think it would be beneficial for people who are interested in adopting the new [inline type import syntax introduced in Typescript 4.5](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/#type-on-import-names). We can introduce a flag, for example `preferInlineTypeImports?: boolean`.

Like @bradzacher mentioned, it's tricky to do a fixer, so we don't need it for now, but at least a way to warn people about violation would be great.

### Fail

```ts
import type { c } from './c';
import { a } from './a'; 
import type { b } from './a';
```

### Pass

```ts
import type { c } from './c';
import { a, type b } from './a';
```

---

Let me know what you think about this. If you give me a green light, I think if I can try working on this feature.

**Comment by bradzacher** [comment]
As I mentioned in my comment here: https://github.com/typescript-eslint/typescript-eslint/issues/3950#issuecomment-978092206
I don't think this rule should be in the business of condensing imports when they are correctly "typed".
Instead that would be the concern of a rule like `no-duplicate-imports` (or even better would be `import/no-duplicates` - #4238).

EG this code:
```ts
import type { cType } from './c';
import { aValue } from './a'; 
import type { bType } from './a';
```
Or this code:
```ts
import type { cType } from './c';
import { aValue, type bType } from './a'; 
```
Would never error in this rule - regardless of config - because all the types are imported via type imports.

----

However, I do think that this lint rule's **fixer** could very much be concerned with fixing to "condensed" imports when configured to do so.
I'd accept option like `fixStyle` which works like this
- `separate-type-imports`
    - always fix type imports to standalone `import type { ... }` statements.
    - reuse existing `import type { ... }` if one exists for the module.
    - This is the current behaviour, and would be the default for the option (or else it'd be a breaking change).
- `inline-type-imports`
    - reuse existing `import type { ... }` if one exists for the module, if one exists.
    - otherwise - always fix type imports to inline `import { type x, ... }`.
- `mixed` - combine the above:
    - reuse existing `import type { ... }` if one exists for the module, if one exists.
    - if there are no value imports for a module, prefer `import type { ... }`.
    - if there are value imports for a module, prefer `import { type x, ... }`.

Fixed output example:
```ts
import type { dType1 } from './d';
import { dType2 } from './d';
import { cType } from './c';
import { aValue, bType } from './a'; 

// `separate-type-imports` fix to
import type { dType1, dType2 } from './d';
import type { cType } from './c';
import type { bType } from './a';
import { aValue } from './a'; 

// `inline-type-imports` fix to
import type { dType1, dType2 } from './d';
import { type cType } from './c';
import { aValue, type bType } from './a'; 

// `mixed` fix to
import type { dType1, dType2 } from './d';
import type { cType } from './c';
import { aValue, type bType } from './a'; 
```

**Comment by akphi** [comment]
Yes, totally agree with having this in `no-duplicate-imports` instead. Do you accept a PR to add a config to that rule? Something like `ignoreTypeImports?: boolean`. I'm honestly not sure of the naming here.

Regarding the fixer, like you said, it could be a little complicated, so I might not pursue it here. But I want to bring to your attention, the `inline-type-imports` fix might not work well. Please see the discussion here - https://github.com/microsoft/TypeScript/issues/47118

Basically

```ts
import { type cType } from './c';
```

will cause a Typescript compilation error when `importsNotUsedAsValues: true` if there is not `non-type` imports.

**Comment by bradzacher** [comment]
> Yes, totally agree with having this in no-duplicate-imports instead. Do you accept a PR to add a config to that rule?

I don't believe we should be maintaining our extension rule, `no-duplicate-imports` - #4238.
There's no reason for us to do so when there is a much better version of the rule in the ecosystem which already supports things - `import/no-duplicates`.

I would like to remove our extension rule in the next major, so I don't hugely want us to add more functionality into it for now.

**Comment by akphi** [comment]
@bradzacher that's a very fair point! I totally forgot about this base rule. I will close this issue now.

**Comment by bradzacher** [comment]
I'll leave this open as a tracking issue in case someone wants to implement the changes to the fixer.

**Comment by gajus** [comment]
You can now achieve this with [`canonical/prefer-inline-type-import`](https://github.com/gajus/eslint-plugin-canonical#eslint-plugin-canonical-rules-prefer-inline-type-import).

**Comment by snewcomer** [comment]
I believe this is still something we want in this repo even with the additions to the canonical?  If so, I've started working on a fix.

**Comment by Josh-Cena** [comment]
@snewcomer It's an ESLint parser error. Did you bump your TS-ESLint parser version?

**Comment by snewcomer** [comment]
@Josh-Cena 👋.  I believe so.  Here is the WIP PR.

https://github.com/emberjs/ember.js/pull/20101

<img width="375" alt="Screen Shot 2022-05-26 at 9 21 58 PM" src="https://user-images.githubusercontent.com/7374640/170615933-6050aa18-0426-4d0b-9799-1401cf9dcfef.png">

**Comment by Josh-Cena** [comment]
Errr... Because you are actually missing a comma...

**Comment by snewcomer** [comment]
I did miss that.  Thank you.  However, still the same issues. Error isn't too revealing.

<img width="516" alt="Screen Shot 2022-05-26 at 10 05 58 PM" src="https://user-images.githubusercontent.com/7374640/170620990-12cedc05-0b80-45bf-a6b2-0777b70eb3bf.png">

**Comment by Josh-Cena** [comment]
Mmm, I hope your VS Code version is the latest? From your previous screenshot it does look like the `type` is correctly highlighted, though. Does running the `tsc` command work?

**Comment by blingerson** [comment]
Is there any way to disable the fixer for consistent-type-imports in the meantime? The fixer just ends up causing hundreds of new issues when used in conjunction with import/no-duplicates.

**Comment by snewcomer** [comment]
@blingerson Could you detail some of the issues you found and perhaps we can fix them?  Even better in a new issue!

**Comment by bradzacher** [comment]
https://github.com/import-js/eslint-plugin-import/issues/2469 has been merged but not yet released.

**Comment by snewcomer** [comment]
@bradzacher Amazing!  

Just to note, we have 2/3 left to get proper inline imports. This PR is next I believe.

https://github.com/import-js/eslint-plugin-import/pull/2475
